segment.go

```diff
@@ -0,0 +1,86 @@
+// mautrix-whatsapp - A Matrix-WhatsApp puppeting bridge.
+// Copyright (C) 2022 Tulir Asokan, Sumner Evans
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+
+	log "maunium.net/go/maulogger/v2"
+	"maunium.net/go/mautrix/id"
+)
+
+var SEGMENT_URL string = "https://api.segment.io/v1/track"
+
+type Segment struct {
+	segmentKey string
+	log        log.Logger
+	client     *http.Client
+}
+
+func NewSegment(segmentKey string, parentLogger log.Logger) *Segment {
+	return &Segment{
+		segmentKey: segmentKey,
+		log:        parentLogger.Sub("Segment"),
+		client:     &http.Client{},
+	}
+}
+
+func (segment *Segment) Track(userID id.UserID, event string, properties ...map[string]interface{}) {
+	if segment.segmentKey == "" {
+		return
+	}
+	if len(properties) > 1 {
+		segment.log.Fatalf("Track should be called with at most one property map")
+	}
+
+	go (func() error {
+		props := map[string]interface{}{}
+		if len(properties) > 0 {
+			props = properties[0]
+		}
```

Comment:
Long inline functions aren't very nice, this should probably be moved to a separate private (lowercase) function

Reply:
I kept a small part of it (for wrangling the properties and handling errors), but extracted out the meat of it